proposals/NNNN-ComparisonReform.md

```diff
@@ -0,0 +1,531 @@
+* Case values of `inOrder`, `same`, `outOfOrder`
+
+The choice of case names is non-trivial because the enum shows up in different
+contexts where different names makes more sense. Effectively, one needs to keep
+in mind that the "default" sort order is ascending to map between the concept
+of "before" and "less".
+
+The before/after naming to provide the most intuitive model for custom sorts --
+referring to `ascending` or `less` is confusing when trying to implement a
+descending ordering. Similarly the inOrder/outOfOrder naming was too indirect
+-- it's more natural to just say where to put the element. If the enum should
+focus on the sorting case, calling it `SortOrder` would help to emphasize this
+fact.
+
+This proposal elects to leave the existing Foundation name in-place. The
+primary motivation for this is that use of the `compare` function will be
+relatively rare. It is expected that in most cases users will continue to make
+use of `==` or `<`, returning boolean values (the main exception to this will
+be in use of the parameterized `String` comparisons). As such, the source
+compatibility consequences of introducing naming changes to an existing type
+seems of insufficient benefit.
+
+The method `compare(_:)` does not fully comport with the API naming guidelines.
+However, it is firmly established with current usage in Objective-C APIs, will
+be fairly rarely seen/used (users will usually prefer `<`, `==` etc), and
+alternatives considered, for example `compared(to:)`, were not a significant
+improvement.
+
+### Add Overloads for `(T, T) -> ComparisonResult`
+
+It would be slightly more ergonomic to work with ComparisonResult if existing
+methods that took an ordering predicate also had an overload for `(T, T) ->
+ComparisonResult`. As it stands, a case-insensitive sort must be written as
+follows:
+
+```swift
+myStrings.sort { $0.compare(_ other: $1, case: .insensitive) == .orderedAscending }
+```
+
+With the overload, one could write:
+
+```swift
+myStrings.sort { $0.compare($1, case: .insensitive) }
+```
+
+we decided against providing these overloads because: 
+
+* The existing algorithms in the standard library can't benefit from them (only binary comparisons).
+* They bloat up the standard library (and any library which intends to match our API guidelines).
+* They potentially introduce confusion over "which" comparison overload to use.
+
+And because we can change our mind later without concern for source or ABI
+stability, as these overloads would be additive.
+
+## Future Work
+
+This section covers some topics which were briefly considered, but were
+identified as reasonable and possible to defer to future releases. Specifically
+they should be backwards compatible to introduce even after ABI stability. Two
+paths that are worth exploring:
+
+
+### Ergonomic Generalized Comparison for Keyed Containers 
```

Comment:
I'm hoping some sort of "anonymous protocol conformance" feature could solve this.